### PR TITLE
Merge in main

### PR DESCRIPTION
**Pull request by TommasoAzz**
Pull request per fare merging del branch.
Prima di fare merging direi di capire cosa è stato rotto. Provo a segnare nei commenti i commit che penso siano problematici.
Ovviamente saranno in ordine

### CODE REVIEW

**Comment by TommasoAzz** [comment]
[f601a2841cd65dd1637d67004c47d2f603076c15](https://github.com/jjocram/expanded-mnist/commit/f601a2841cd65dd1637d67004c47d2f603076c15)
Sicuro funzionante perché commit da cui @jjocram ha ottenuto buoni risultati e i risultati erano questi:
photo_2022-01-25 15 11 18

**Comment by jjocram** [comment]
Questo è il mio ultimo commit: e41335d4c1d42b9f7ee5bd4b9b98ca00bb6c85cb

**Comment by TommasoAzz** [comment]
- [88bd71483dbeb970963053d9ad9f75d9fc317583](https://github.com/jjocram/expanded-mnist/commit/88bd71483dbeb970963053d9ad9f75d9fc317583) in yolo/utils/metrics.py usa i .item() quando viene invocata **compute_accuracy_metrics** e la funzione **_accuracy_value** ritorna float. Non vorrei che il casting facesse perdere info...sarebbe drammatico.

**Comment by TommasoAzz** [comment]
> Questo è il mio ultimo commit: [e41335d](https://github.com/jjocram/expanded-mnist/commit/e41335d4c1d42b9f7ee5bd4b9b98ca00bb6c85cb)

Okay, nice to know. Quello che ho scritto che va è quello che dava questi risultati qua:
photo_2022-01-25 15 17 00

File: yolo/utils/checkpoints.py [hunk 1]
```diff
@@ -1,24 +1,57 @@
...
+    """
+    checkpoint = torch.load(input_file_name, map_location=utils.DEVICE)
     model.load_state_dict(checkpoint["state_dict"])
     optimizer.load_state_dict(checkpoint["optimizer"])
 
-    # If we don't do this then it will just have learning rate of old checkpoint
-    # and it will lead to many hours of debugging \:
-    for param_group in optimizer.param_groups:
-        param_group["lr"] = lr
+    if learning_rate > 0.0:
```
Comment (jjocram) [inline yolo/utils/checkpoints.py]: Perché questo if?

File: yolo/utils/__init__.py [hunk 2]
```diff
@@ -1,27 +1,7 @@
...
 from .checkpoints import save_checkpoint, load_checkpoint
 from .intersection_over_union import iou_width_height, intersection_over_union
 from .mean_average_precision import mean_average_precision
-from .metrics import check_class_accuracy
+from .metrics import compute_accuracy_metrics
 from .non_maximum_suppression import non_max_suppression
 from .plot import plot_couple_examples, plot_image
-
-
-def get_mean_std(loader):
```
Comment (jjocram) [inline yolo/utils/__init__.py]: Questo metodo (get_mean_std) è stato rimosso perché non veniva usato o perché è stato spostato?

Comment (TommasoAzz) [reply yolo/utils/__init__.py]: mai usato

File: yolo/utils/checkpoints.py (hunk 1 above)
Comment (TommasoAzz) [reply yolo/utils/checkpoints.py]: Perché nessuno mi setta 0.0 come learning rate (non apprenderebbe nulla). Infatti è il valore di default. Quindi dico "se non mi hai settato il learning rate tanto vale che mi metta ad aggiornarlo"

**Comment by TommasoAzz** [comment]
- [a40c97c62240f2c05fecd584ba2bc52918cadefc](https://github.com/jjocram/expanded-mnist/commit/a40c97c62240f2c05fecd584ba2bc52918cadefc) qui sostituisco una lista di tensori in tensore...è la modifica più sostanziale che ho fatto...e magari è quella che rompe qualcosa. è vero anche che è una metrica, non la rete stessa

File: yolo/utils/metrics.py [hunk 3]
```diff
@@ -1,36 +1,68 @@
...
+        for i in range(len(out)):
             y[i] = y[i].to(utils.DEVICE)
-            obj = y[i][..., 0] == 1  # in paper this is Iobj_i
-            noobj = y[i][..., 0] == 0  # in paper this is Iobj_i
+            # The following obj and noobj are the 1^obj_i and 1^noobj_i functions in the YOLO loss function.
+            obj = y[i][..., 0] == 1
+            noobj = y[i][..., 0] == 0
 
-            correct_class += torch.sum(
+            correct_class_predictions += torch.sum(
```
[the remark this answers is not in the document]
Comment (jjocram) [reply yolo/utils/metrics.py]: E più sotto a volte i .item() corrispondono e a volte no

File: yolo/utils/metrics.py [hunk 4]
```diff
@@ -1,36 +1,68 @@
...
+
+            obj_predictions = torch.sigmoid(out[i][..., 0]) > threshold
+
+            correct_obj += torch.sum(obj_predictions[obj] == y[i][..., 0][obj]).item()
+            tot_obj += torch.sum(obj).item()
+
+            correct_noobj += torch.sum(obj_predictions[noobj] == y[i][..., 0][noobj]).item()
+            tot_noobj += torch.sum(noobj).item()
+
+    class_prediction_accuracy = _accuracy_value(correct_class_predictions, tot_class_predictions) / 100
```
Comment (jjocram) [inline yolo/utils/metrics.py]: Perché questo `/ 100`?

**Comment by TommasoAzz** [comment]
- [d2fc1ca45eec89d99c96cf7e3f969a832a1e2512](https://github.com/jjocram/expanded-mnist/commit/d2fc1ca45eec89d99c96cf7e3f969a832a1e2512) ecco QUI uso `torch.no_grad()` quindi è da dare un occhio
- [56a8a15c1d2d560dbc165131454c1c76c3e4d826](https://github.com/jjocram/expanded-mnist/commit/56a8a15c1d2d560dbc165131454c1c76c3e4d826) è da guardare tutto perché cambio il sistema di training

**Comment by TommasoAzz** [comment]
- [1df7df134a3df6ee083efc945f3c91d9ffde93e1](https://github.com/jjocram/expanded-mnist/commit/1df7df134a3df6ee083efc945f3c91d9ffde93e1) rimuovo vecchi notebook
- [621ab6f7b2d9ab96d8945a09e753e1249bb3975d](https://github.com/jjocram/expanded-mnist/commit/621ab6f7b2d9ab96d8945a09e753e1249bb3975d) aggiungo parametri di configurazione per train.py ma nuovamente metto mani al training, meglio dare un occhio

File: yolo/utils/metrics.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TommasoAzz) [reply yolo/utils/metrics.py]: No, li ho messi io per avere risultati in float e non in tensori di dim 1x1. A sto punto si può per evitare problemi ritornare tensori 1x1. Aggiorno e ti metto qui l'id del commit poi.

[the remark this answers is not in the document]
Comment (TommasoAzz) [reply yolo/utils/metrics.py]: [5d324e93cefc6ace78ad143552d8961267ae6250](https://github.com/jjocram/expanded-mnist/commit/5d324e93cefc6ace78ad143552d8961267ae6250) ecco qua. lancio su slurm e vedo se gira

**Comment by TommasoAzz** [comment]
> @TommasoAzz perché qui è rimasta la notazione `{}".format`

è voluta per andare multiline

**Comment by TommasoAzz** [comment]
@jjocram controlla [9e1aa0b69a736cc2513ce5f30aba7309e50a06cd](https://github.com/jjocram/expanded-mnist/commit/9e1aa0b69a736cc2513ce5f30aba7309e50a06cd)

**Comment by jjocram** [comment]
> @jjocram controlla [9e1aa0b69a736cc2513ce5f30aba7309e50a06cd](https://github.com/jjocram/expanded-mnist/commit/9e1aa0b69a736cc2513ce5f30aba7309e50a06cd) che lo lancio su SLURM

Mi sembra vada bene, prova a lanciarlo
